src/chunk_adaptive.c

```diff
@@ -0,0 +1,551 @@
+	 */
+	return estimate_effective_memory_cache_size() / DEFAULT_CONCURRENT_CHUNK_USAGE;
+}
+
+/*
+ * Use a heap scan to find the min and max of a given column of a chunk. This
+ * could be a rather costly operation. Should figure out how to keep min-max
+ * stats cached.
+ *
+ * Returns true iff min and max is found.
+ */
+static bool
+minmax_heapscan(Relation rel, Oid atttype, AttrNumber attnum, Datum minmax[2])
+{
+	HeapScanDesc scan;
+	HeapTuple	tuple;
+	TypeCacheEntry *tce;
+	bool		minmaxnull[2] = {true};
+
+	/* Lookup the tuple comparison function from the type cache */
+	tce = lookup_type_cache(atttype, TYPECACHE_CMP_PROC | TYPECACHE_CMP_PROC_FINFO);
+
+	if (NULL == tce || !OidIsValid(tce->cmp_proc))
+		elog(ERROR, "no comparison function for type %u", atttype);
+
+	scan = heap_beginscan(rel, GetTransactionSnapshot(), 0, NULL);
+
+	while ((tuple = heap_getnext(scan, ForwardScanDirection)) != NULL)
+	{
+		bool		isnull;
+		Datum		value = heap_getattr(tuple, attnum, RelationGetDescr(rel), &isnull);
+
+		/* Check for new min */
+		if (minmaxnull[0] || DatumGetInt32(FunctionCall2(&tce->cmp_proc_finfo, value, minmax[0])) < 0)
+		{
+			minmaxnull[0] = false;
+			minmax[0] = value;
+		}
+
+		/* Check for new max */
+		if (minmaxnull[1] || DatumGetInt32(FunctionCall2(&tce->cmp_proc_finfo, value, minmax[1])) > 0)
+		{
+			minmaxnull[1] = false;
+			minmax[1] = value;
+		}
+	}
+
+	heap_endscan(scan);
+
+	return !minmaxnull[0] && !minmaxnull[1];
+}
+
+/*
+ * Use an index scan to find the min and max of a given column of a chunk.
+ *
+ * Returns true iff min and max is found.
+ */
+static bool
+minmax_indexscan(Relation rel, Relation idxrel, AttrNumber attnum, Datum minmax[2])
+{
+	IndexScanDesc scan = index_beginscan(rel, idxrel, GetTransactionSnapshot(), 0, 0);
+	HeapTuple	tuple;
+	bool		isnull;
+	int			n = 0;
+
+	tuple = index_getnext(scan, BackwardScanDirection);
+
+	if (HeapTupleIsValid(tuple))
+		minmax[n++] = heap_getattr(tuple, attnum, RelationGetDescr(rel), &isnull);
+
+	index_rescan(scan, NULL, 0, NULL, 0);
+	tuple = index_getnext(scan, ForwardScanDirection);
+
+	if (HeapTupleIsValid(tuple))
+		minmax[n++] = heap_getattr(tuple, attnum, RelationGetDescr(rel), &isnull);
+
+	index_endscan(scan);
+
+	return n == 2;
+}
+
+/*
+ * Get the min and max value for a given column of a chunk.
+ *
+ * Returns true iff min and max is found, otherwise false.
+ */
+static bool
+chunk_get_minmax(Oid relid, Oid atttype, AttrNumber attnum, Datum minmax[2])
+{
+	Relation	rel = heap_open(relid, AccessShareLock);
+	List	   *indexlist = RelationGetIndexList(rel);
+	ListCell   *lc;
+	bool		found = false;
+
+	foreach(lc, indexlist)
+	{
+		Relation	idxrel;
+
+		idxrel = index_open(lfirst_oid(lc), AccessShareLock);
+
+		if (idxrel->rd_att->attrs[0]->attnum == attnum)
+			found = minmax_indexscan(rel, idxrel, attnum, minmax);
+
+		index_close(idxrel, AccessShareLock);
+
+		if (found)
+			break;
+	}
+
+	if (!found)
+		found = minmax_heapscan(rel, atttype, attnum, minmax);
```

Comment:
I would not do a heapscan. Too costly. I think the logic should be: Do an indexscan, and if there is no appropriate index, abandon operation with a WARNING in the log.

Reply:
I assume you mean fall back to current chunk size when you say abandon? We cannot fail the insert.